dune-precice/dune-precice/couplinginterface.hh

```diff
@@ -0,0 +1,264 @@
+  void CouplingInterface<dim, VectorType, ParameterClass>::send_blockvector_data(const VectorType &dune_data)
+  {
+    Utilities::copy_from_dune_to_precice<dim, VectorType>(dune_data, write_data_, is_coupling_boundary_);
+    precice.writeBlockVectorData(write_data_id_, n_interface_nodes_, interface_nodes_ids_.data(), write_data_.data());
+  }
+      
+  template <int dim, typename VectorType, typename ParameterClass>
+  double CouplingInterface<dim, VectorType, ParameterClass>::advance(double computed_timestep_length)
+  {
+    return precice.advance(computed_timestep_length);
+  }
+  
+  template <int dim, typename VectorType, typename ParameterClass>
+  bool CouplingInterface<dim, VectorType, ParameterClass>::is_coupling_ongoing() const
+  {
+    return precice.isCouplingOngoing();
+  }
+  
+  template <int dim, typename VectorType, typename ParameterClass>
+  bool CouplingInterface<dim, VectorType, ParameterClass>::is_time_window_complete() const
+  {
+    return precice.isTimeWindowComplete();
+  }
+  
+  template <int dim, typename VectorType, typename ParameterClass>
+  void CouplingInterface<dim, VectorType, ParameterClass>::finalize()
+  {
+    precice.finalize();
+  }
+  
+  template <int dim, typename VectorType, typename ParameterClass>
+  bool CouplingInterface<dim, VectorType, ParameterClass>::is_save_required() const
+  {
+    return precice.isActionRequired(precice::constants::actionWriteIterationCheckpoint());
+  }
+  
+  template <int dim, typename VectorType, typename ParameterClass>
+  void CouplingInterface<dim, VectorType, ParameterClass>::mark_save_fullfilled() {
+    precice.markActionFulfilled(precice::constants::actionWriteIterationCheckpoint());
+  }
+  
+  template <int dim, typename VectorType, typename ParameterClass>
+  bool CouplingInterface<dim, VectorType, ParameterClass>::is_load_required() const
+  {
+    return precice.isActionRequired(precice::constants::actionReadIterationCheckpoint());
+  }
+  
+  template <int dim, typename VectorType, typename ParameterClass>
+  void CouplingInterface<dim, VectorType, ParameterClass>::mark_load_fullfilled()
+  {
+    precice.markActionFulfilled(precice::constants::actionReadIterationCheckpoint());
+  }
+  
+  template <int dim, typename VectorType, typename ParameterClass>
+  void CouplingInterface<dim, VectorType, ParameterClass>::save_current_state(
+    const std::vector<VectorType*>& state_quantaties,
+    const double time,
+    const int iter) 
+  {  
+    old_state_data_.resize(state_quantaties.size());
+    
+    for (int i=0; i<state_quantaties.size(); i++) {
+      old_state_data_[i] = *(state_quantaties[i]);
+    }
+
+    old_time_ = time;
+    old_iter_ = iter; 
+  }
+
+  template <int dim, typename VectorType, typename ParameterClass>
+  void CouplingInterface<dim, VectorType, ParameterClass>::reload_old_state(
+    std::vector<VectorType*>& state_quantaties,
+    double &time,
+    int &iter)
+  {
+    for (int i=0; i<state_quantaties.size(); i++) {
```

Comment:
```suggestion
    assert( state_quantaties.size() == old_state_data_.size() );
    for (int i=0; i<state_quantaties.size(); i++) {
```
Should we add an assertion here or is it guaranteed that the size of `state_quantaties` has not changed?

Reply:
I guess it won't hurt having one.